policy_parser.py

```diff
@@ -14,35 +14,107 @@ def __init__(self, msg=None, policy=None):
-            # Convert each allow_incoming_clause into one or more
-            # Calico Rule objects.
-            if allow_incoming_clause:
-                # Rule exists - parse it.
-                r = self._allow_incoming_to_rules(allow_incoming_clause)
-            else:
-                # An empty rule means allow all traffic.
-                r = [Rule(action="allow")]
-            rules.extend(r)
+        ingress_rules = self.policy["spec"].get("ingress")
+        if ingress_rules:
+            _log.debug("Got %d ingress rules: translating to Calico format",
+                       len(ingress_rules))
+            for ingress_rule in ingress_rules:
+                _log.debug("Processing ingress rule %s", ingress_rule)
+                if ingress_rule:
+                    # Convert ingress rule into Calico Rules.
+                    _log.debug("Adding rule %s", ingress_rule)
+                    rules.extend(self._allow_incoming_to_rules(ingress_rule))
+                else:
+                    # An empty rule means allow all traffic.
+                    _log.debug("Empty rule => allow all; skipping rest")
+                    rules.append(Rule(action="allow"))
+                    break
 
         _log.debug("Calculated total set of rules: %s", rules)
         return rules
 
+    def _calculate_selectors(self, label_selector, key_format="%s"):
+        """
+        Generate Calico datamodel selectors for a Kubernetes LabelSelector.
+
+        Returns a list of selectors in the Calico datamodel format.
+        Throws PolicyError if the LabelSelector is invalid or unsupported.
+        """
+        # A null LabelSelector matches no objects.
+        calico_selectors = []
+
+        # matchLabels is a map key => value, it means match if (label[key] ==
+        # value) for all keys.
+        if "matchLabels" in label_selector:
+            labels = label_selector["matchLabels"]
+            calico_selectors += [
+                "%s == '%s'" % (key_format % k, v) for k, v in labels.iteritems()
+            ]
+        
+        # matchExpressions is a list of in/notin/exists/doesnotexist tests.
+        if "matchExpressions" in label_selector:
+            for expression in label_selector["matchExpressions"]:
+                key = key_format % expression["key"]
+                operator = expression["operator"]
+                values = expression.get("values", [])
+                value_list = ", ".join(["\"%s\"" % v for v in values])
+                if operator == "In":
+                    calico_selectors.append("%s in { %s }" % (key, value_list))
+                elif operator == "NotIn":
+                    calico_selectors.append("%s not in { %s }" % (key, value_list))
+                elif operator == "Exists":
+                    calico_selectors.append("has(%s)" % key)
+                elif operator == "DoesNotExist":
+                    calico_selectors.append("! has(%s)" % key)
```

Comment:
I went with the exact forms in the etcd-data-model doc.  Really, I suspect it strips all spaces.